### ISSUE TITLE
[BUG] Soci snapshotter only lazy loads one image

### ISSUE THREAD

**Issue by Kern--** [awslabs/soci-snapshotter#73]
**Describe the bug**
tl;dr the snapshotter falls back to ahead-of-time pulls after the first image pull because fs.filesystem caches the first SOCI index it sees and tries to use that data for all other image pulls.

[fs.filesystem](https://github.com/awslabs/soci-snapshotter/blob/b32d03ad7406e66acfbae3d1e6bf15863283a18d/fs/fs.go#L185) is responsible for reading labels and setting up FUSE mounts for each layer of an image when pulled using a SOCI index. When the first layer of an image is passed to the filesystem, the filesystem does a [one time SOCI index fetch](https://github.com/awslabs/soci-snapshotter/blob/b32d03ad7406e66acfbae3d1e6bf15863283a18d/fs/fs.go#L206) to download the SOCI index from the remote registry. This onetime setup also assigns the [index and a mapping of layer -> ztoc descriptor](https://github.com/awslabs/soci-snapshotter/blob/b32d03ad7406e66acfbae3d1e6bf15863283a18d/fs/fs.go#L198) in the filesystem itself so we have that information ready for the next layer without an additional network call.

The problem is that the [fs.filesystem] is not unique for each image, but rather there is a single instance for the entire snapshotter. As a result, the first image that gets pulled using a soci index will load that index and it's layers into the filesystem data structure and future image pulls will not change this data. For each layer in the second image, a lookup of the [ztoc digest](https://github.com/awslabs/soci-snapshotter/blob/b32d03ad7406e66acfbae3d1e6bf15863283a18d/fs/fs.go#L304) will fail because it is not in the first image's layer -> ztoc map which causes Resolve to fail as it's trying to resolve an empty descriptor. The resolution error bubbles up to the snapshotter service which falls back to ahead-of-time pulling with overlayfs. 

We still want to hold on to the index and layer -> ztoc map, but we need to be able to hold on to multiple instances of these at a time. Wrapping those two components up into a `IndexInfo` struct (ugh. naming is hard) and then storing those in the filesystem in a [sync.Map](https://pkg.go.dev/sync#Map) of `indexDigest string` -> `IndexInfo` seems like a good way to handle this.

**Steps To Reproduce**
Start the snapshotter and `soci rpull` any two images for which there are SOCI indices. 

**Expected behavior**
Both images should be pulled lazily using their respective 

**Configuration (please complete the following information):**
 - OS: AmazonLinux 2
 - Snapshotter Version: soci-snapshotter-grpc b32d03a b32d03ad7406e66acfbae3d1e6bf15863283a18d
 - Containerd Version: 1.6.0

**Additional context**
